# 20440_TARA


## Overview:

This repository is a home for Rashi Jeeda and Victoria Chen's final project for MIT 20.440 Analysis of Biological Networks. The project is titled "Uncovering drivers of ocean microbiome diversity". Our goal is to use metadata and microbial taxonomic classifications generated from Tara Oceans expeditions to understand the principal drivers of microbiome diversity in oceans. This repository contains data and code for analysis of the Tara Oceans microbiome dataset.

## Data:

The data was generated by the Tara Ocean expedition team. Ocean water samples were collected in 68 locations across the globe to create a ocean microbial gene catalog. Metadata was collected for all ocean sampling sites including ocean name, latitude, longitude, date, time, depth, temperature, and more. Metagenomic merged Illumina reads were extracted for all samples and 16S rRNA was sequenced for taxonomic classification of all microbes found in each samples. 16S rRNA reads were mapped to 16S reference sequences from the widely-used SILVA database and clustered into 97% operational taxonomic units (OTUs), which provide a measure of relative abundances of various taxa found in samples. The pre-processed OTU count table was provided.

Citation: "Structure and function of the global ocean microbiome" by Sunagawa, Coelho, Chaffron, et al., Science, 2015.


## Required downloads:
Files too large to be uploaded to this repository can be downloaded from the Tara Oceans companion website at [ocean-microbiome.embl.de/companion.html](http://ocean-microbiome.embl.de/companion.html). The files required for this analysis are:

```Companion Website Tables W1-W8```, a spreadsheet file containing multiple types of metadata

```annotated 16S OTU count table```, containing OTU counts summarized at different taxonomic levels for prokaryote-enriched samples (N=139)

```eggNOG OG profile (prokaryotic only)```, a table of eggNOG gene family abundances for prokaryote-enriched samples (N=139)
    

## Folder structure:

```
.
├── data
│   ├── raw: datasets as downloaded from the Tara Oceans paper companion website
│   └── clean: renamed datasets saved in .csv format and outputs of analysis code
├── plots: subplots generated for each figure
│   └── figure_1
│   └── figure_2
│      └── scatter: scatterplots of Random Forest regressors for each type of input data
│            └── OTU
│            └── OG
│            └── phyla
│            └── function
│      └── confusion: confusion matrices of Random Forest classifiers for each type of input data
│            └── OTU
│            └── OG
│            └── phyla
│            └── function
│   └── figure_3
│   └── figure_4
│      └── important_trends: top trends (relative abundance vs environmental variable) of interest
│      └── importances_regr: top feature importances for Random Forest regressors for each type of input data
│            └── OTU
│            └── OG
│            └── phyla
│            └── function
│      └── importances_rfc: top feature importances for Random Forest classifiers for each type of input data
│            └── OTU
│            └── OG
│            └── phyla
│            └── function
│   └── figure_S1
├── plotting_scripts: Jupyter notebooks for analysis and plotting done for each figure
│   └── figure_1_3
│   └── figure_2
│      └── random_forest_models: Jupyter notebooks for running the Random Forest models
│   └── figure_4
│   └── figure_S1
│   └── other: all other plotting scripts, such as exploratory data analysis
└── processing_scripts: scripts used to create clean relative abundance tables with metadata
 ```
 
## Installation:

The following packages should be installed for running both the plotting scripts and processing scripts (also found in requirements.txt):

```
python==3.8.11 

biom_format==2.1.15

bokeh==2.3.3

cmocean==3.0.3

holoviews==1.14.6

iqplot==0.2.3

matplotlib==3.4.2

numpy==1.22.4

pandas==1.3.3

scikit_bio==0.5.8

scikit_learn==0.24.2

scipy==1.7.1

seaborn==0.11.2
```
