{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b01e8244-18bf-495e-b2ae-c864a28def55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.3.min.js\": \"dM3QQsP+wXdHg42wTqW85BjZQdLNNIXqlPw/BgKoExPmTG7ZLML4EGqLMfqHT6ON\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.3.min.js\": \"8x57I4YuIfu8XyZfFo0XVr2WAT8EK4rh/uDe3wF7YuW2FNUSNEpJbsPaB1nJ2fz2\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.3.min.js\": \"3QTqdz9LyAm2i0sG5XTePsHec3UHWwVsrOL68SYRoAXsafvfAyqtQ+h440+qIBhS\"};\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      if (url in hashes) {\n",
       "        element.crossOrigin = \"anonymous\";\n",
       "        element.integrity = \"sha384-\" + hashes[url];\n",
       "      }\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.3.min.js\"];\n",
       "  var css_urls = [];\n",
       "  \n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (var i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.3.min.js\": \"dM3QQsP+wXdHg42wTqW85BjZQdLNNIXqlPw/BgKoExPmTG7ZLML4EGqLMfqHT6ON\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.3.min.js\": \"8x57I4YuIfu8XyZfFo0XVr2WAT8EK4rh/uDe3wF7YuW2FNUSNEpJbsPaB1nJ2fz2\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.3.min.js\": \"3QTqdz9LyAm2i0sG5XTePsHec3UHWwVsrOL68SYRoAXsafvfAyqtQ+h440+qIBhS\"};\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      if (url in hashes) {\n        element.crossOrigin = \"anonymous\";\n        element.integrity = \"sha384-\" + hashes[url];\n      }\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.3.min.js\"];\n  var css_urls = [];\n  \n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (var i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Imports\n",
    "import os\n",
    "\n",
    "import bokeh\n",
    "import bokeh.plotting\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "bokeh.io.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "81fb9a7b-1df2-4ae8-b421-44804da8e2f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set path to collect data files\n",
    "data_path = \"../data/raw/\"\n",
    "\n",
    "# Import necessary files\n",
    "OG_file = os.path.join(data_path, \"TARA243.OG.profile.release\")\n",
    "COG_file = os.path.join(data_path, \"COG.funccat.txt\")\n",
    "NOG_file = os.path.join(data_path, \"NOG.funccat.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6e5c2ed-1f8c-4f52-8e42-4b6908700f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read each input data file into pandas\n",
    "OG_df = pd.read_csv(OG_file, sep='\\t')\n",
    "COG_df = pd.read_csv(COG_file, sep='\\t', encoding='cp1252', names=['OG', 'category'])\n",
    "NOG_df = pd.read_csv(NOG_file, sep='\\t', encoding='cp1252', names=['OG', 'category'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b67d0f71-7d22-45e4-a7d7-4290e98eb6dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set path to collect data files\n",
    "data_path = \"../data/clean/\"\n",
    "\n",
    "# Import necessary files\n",
    "metadata_file = os.path.join(data_path, \"companion_table_W1.csv\")\n",
    "nutrient_file = os.path.join(data_path, \"nutrient_temp_table.csv\")\n",
    "core_file = os.path.join(data_path, \"core_annos.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3d59c7a5-4984-449a-8d31-d2250629de64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read each input data file into pandas\n",
    "meta_df = pd.read_csv(metadata_file)\n",
    "cond_df = pd.read_csv(nutrient_file)\n",
    "core_df = pd.read_csv(core_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d26c62f-eb75-427f-bfd3-392c88555d66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract sample label, id, and location from metadata\n",
    "id_df = meta_df.iloc[:,[0, 4, 11, 14, 15, 16]]\n",
    "id_df = id_df.rename(columns={'Sample label [TARA_station#_environmental-feature_size-fraction]': 'label', \n",
    "                              'PANGAEA sample identifier': 'pangea_id',\n",
    "                              'Marine pelagic biomes (Longhurst 2007)': 'Biome',\n",
    "                              'Ocean and sea regions (IHO General Sea Areas 1953) [MRGID registered at www.marineregions.com]': 'Region',\n",
    "                              'Marine pelagic biomes  (Longhurst 2007) [MRGID registered at www.marineregions.com] ': 'Province'\n",
    "                             })\n",
    "# Extract temperature and id info for each sample\n",
    "cond_df = cond_df.iloc[:,[0, 2, 3, 4, 5, 6, 7, 8, 9, 10]]\n",
    "cond_df = cond_df.rename(columns={'PANGAEA Sample ID': 'pangea_id', \n",
    "                                  'Mean_Lat*': 'Latitude',\n",
    "                                  'Mean_Long*': 'Longitude',\n",
    "                                  'Mean_Depth [m]*': 'Depth (m)',\n",
    "                                  'Mean_Temperature [deg C]*': 'Temperature (°C)',\n",
    "                                  'Mean_Salinity [PSU]*': 'Salinity (PSU)',\n",
    "                                  'Mean_Oxygen [umol/kg]*': 'Oxygen (µmol/kg)',\n",
    "                                  'Mean_Nitrates[umol/L]*': 'Nitrates (µmol/L)',\n",
    "                                  'NO2 [umol/L]**': 'NO2 (µmol/L)',\n",
    "                                  'PO4 [umol/L]**': 'PO4 (µmol/L)',\n",
    "                                 })\n",
    "# Merge all relevant data into one df\n",
    "id_df = id_df.merge(cond_df, how='left', on='pangea_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7935cd50-636d-431e-8364-392f49c00500",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OG</th>\n",
       "      <th>label</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sum_not_annotated</td>\n",
       "      <td>TARA_100_DCM_0.22-3</td>\n",
       "      <td>7.617148e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COG0001</td>\n",
       "      <td>TARA_100_DCM_0.22-3</td>\n",
       "      <td>9.924748e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COG0002</td>\n",
       "      <td>TARA_100_DCM_0.22-3</td>\n",
       "      <td>9.434435e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COG0003</td>\n",
       "      <td>TARA_100_DCM_0.22-3</td>\n",
       "      <td>2.337477e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COG0004</td>\n",
       "      <td>TARA_100_DCM_0.22-3</td>\n",
       "      <td>2.662103e+04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  OG                label        counts\n",
       "0  sum_not_annotated  TARA_100_DCM_0.22-3  7.617148e+06\n",
       "1            COG0001  TARA_100_DCM_0.22-3  9.924748e+03\n",
       "2            COG0002  TARA_100_DCM_0.22-3  9.434435e+03\n",
       "3            COG0003  TARA_100_DCM_0.22-3  2.337477e+02\n",
       "4            COG0004  TARA_100_DCM_0.22-3  2.662103e+04"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OG_df = pd.melt(OG_df, [\"cog\"], var_name=\"label\", value_name=\"counts\")\n",
    "OG_df = OG_df.rename(columns={\"cog\": \"OG\"})\n",
    "OG_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1957ccfa-8385-4634-a8a2-5488bb36617c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary of OG categories\n",
    "OG_category_dict = {\n",
    "    'A': 'RNA processing and modification',\n",
    "    'B': 'Chromatin structure and dynamics',\n",
    "    'C': 'Energy production and conversion',\n",
    "    'D': 'Cell cycle control and mitosis',\n",
    "    'E': 'Amino acid metabolism and transport',\n",
    "    'F': 'Nucleotide metabolism and transport',\n",
    "    'G': 'Carbohydrate metabolism and transport',\n",
    "    'H': 'Coenzyme metabolism',\n",
    "    'I': 'Lipid metabolism',\n",
    "    'J': 'Translation',\n",
    "    'K': 'Transcription',\n",
    "    'L': 'Replication and repair',\n",
    "    'M': 'Cell wall/membrane/envelope biogenesis',\n",
    "    'N': 'Cell motility',\n",
    "    'O': 'Post-translational modification, protein turnover, chaperone functions',\n",
    "    'P': 'Inorganic ion transport and metabolism',\n",
    "    'Q': 'Secondary structure',\n",
    "    'T': 'Signal transduction',\n",
    "    'U': 'Intracellular trafficing and secretion',\n",
    "    'V': 'Defense mechanisms',\n",
    "    'W': 'Extracellular structures',\n",
    "    'X': 'Mobilome: prophages, transposons',\n",
    "    'Y': 'Nuclear structure',\n",
    "    'Z': 'Cytoskeleton',\n",
    "    'R': 'General functional prediction only',\n",
    "    'S': 'Function unknown'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "96113e13-4800-4dad-a3aa-3ee646cad011",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use dictionary to convert letter abbreviations to categories\n",
    "COG_df['category'] = COG_df['category'].astype(str).str[0] # Only taking first function, not sure yet how to handle multifunction genes\n",
    "NOG_df['category'] = NOG_df['category'].astype(str).str[0]\n",
    "COG_df = COG_df.replace({\"category\": OG_category_dict})\n",
    "NOG_df = NOG_df.replace({\"category\": OG_category_dict})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "468a67e8-7ed3-4511-819f-027662112a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_category_df = pd.concat([COG_df, NOG_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e6c9d414-545c-4493-b1f4-64118abc73cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OG</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COG0393</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COG4055</td>\n",
       "      <td>Coenzyme metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COG1943</td>\n",
       "      <td>Replication and repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COG0861</td>\n",
       "      <td>Inorganic ion transport and metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COG3010</td>\n",
       "      <td>Carbohydrate metabolism and transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84000</th>\n",
       "      <td>NOG205709</td>\n",
       "      <td>Post-translational modification, protein turno...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84001</th>\n",
       "      <td>NOG240538</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84002</th>\n",
       "      <td>NOG145935</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84003</th>\n",
       "      <td>NOG145934</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84004</th>\n",
       "      <td>NOG145936</td>\n",
       "      <td>Function unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88878 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              OG                                           category\n",
       "0        COG0393                                   Function unknown\n",
       "1        COG4055                                Coenzyme metabolism\n",
       "2        COG1943                             Replication and repair\n",
       "3        COG0861             Inorganic ion transport and metabolism\n",
       "4        COG3010              Carbohydrate metabolism and transport\n",
       "...          ...                                                ...\n",
       "84000  NOG205709  Post-translational modification, protein turno...\n",
       "84001  NOG240538                                   Function unknown\n",
       "84002  NOG145935                                   Function unknown\n",
       "84003  NOG145934                                   Function unknown\n",
       "84004  NOG145936                                   Function unknown\n",
       "\n",
       "[88878 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OG_category_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7af488a6-a55f-4b84-a2ba-5849212bd959",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_df = OG_df.merge(OG_category_df, how=\"left\", on=\"OG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5235a655-2ebf-4aa9-9434-22e807a61a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add core annotations \n",
    "core_df = core_df.rename(columns={\"COG\": \"OG\"})\n",
    "OG_df = OG_df.merge(core_df, how=\"left\", on=\"OG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0ae03050-83c3-4667-b363-9398862e4955",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Not sure whether to drop the NOGs that don't map to anything or do function unknown\n",
    "# I think this is closer to the paper's figure\n",
    "# OG_df = OG_df.replace(np.nan, 'Function unknown')\n",
    "OG_df = OG_df[OG_df['category'].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db3a7e5f-17e0-4259-b63f-285bc8d622ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop all gene categories where we don't know the function\n",
    "OG_df = OG_df.loc[OG_df[\"Ocean core\"] == False]\n",
    "OG_df = OG_df.loc[OG_df[\"category\"] != \"Function unknown\"]\n",
    "OG_df = OG_df.loc[OG_df[\"category\"] != \"General functional prediction only\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "85806e90-55c1-4f9f-84a2-acaf41f548d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "abundances_list = []\n",
    "sample_list = []\n",
    "function_list = []\n",
    "# Calculate relative abundances for each sample\n",
    "for sample in OG_df.label.unique():\n",
    "    total = OG_df.loc[OG_df['label'] == sample]['counts'].sum()\n",
    "    abundances = OG_df.loc[OG_df['label'] == sample].groupby('category')['counts'].sum()/total\n",
    "    abundances_list.extend(abundances.values)\n",
    "    sample_list.extend([sample] * len(abundances))\n",
    "    function_list.extend(list(abundances.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cb271351-1bea-493c-a362-c7fe4bd8f801",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_relative_df = pd.DataFrame({\n",
    "    'abundance': abundances_list,\n",
    "    'label': sample_list,\n",
    "    'Function': function_list\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "18cba545-fb67-48f0-995e-45b4a00be1b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_relative_df = OG_relative_df.pivot(index=\"label\", columns=[\"Function\"], values=\"abundance\").fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "79d41f62-f4a2-4e2d-8884-979f5e578bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_relative_df = OG_relative_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "60e997ed-e34c-4e9e-9845-e907c9975adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_relative_df = OG_relative_df.merge(id_df, how=\"left\", on=\"label\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6f4799f5-95cb-4695-a6a0-4360aa3e5b28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>Amino acid metabolism and transport</th>\n",
       "      <th>Carbohydrate metabolism and transport</th>\n",
       "      <th>Cell cycle control and mitosis</th>\n",
       "      <th>Cell motility</th>\n",
       "      <th>Cell wall/membrane/envelope biogenesis</th>\n",
       "      <th>Chromatin structure and dynamics</th>\n",
       "      <th>Coenzyme metabolism</th>\n",
       "      <th>Cytoskeleton</th>\n",
       "      <th>Defense mechanisms</th>\n",
       "      <th>...</th>\n",
       "      <th>Province</th>\n",
       "      <th>Latitude</th>\n",
       "      <th>Longitude</th>\n",
       "      <th>Depth (m)</th>\n",
       "      <th>Temperature (°C)</th>\n",
       "      <th>Salinity (PSU)</th>\n",
       "      <th>Oxygen (µmol/kg)</th>\n",
       "      <th>Nitrates (µmol/L)</th>\n",
       "      <th>NO2 (µmol/L)</th>\n",
       "      <th>PO4 (µmol/L)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TARA_004_DCM_0.22-1.6</td>\n",
       "      <td>0.083391</td>\n",
       "      <td>0.062391</td>\n",
       "      <td>0.004614</td>\n",
       "      <td>0.020305</td>\n",
       "      <td>0.019107</td>\n",
       "      <td>0.011587</td>\n",
       "      <td>0.076737</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.010855</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>36.57</td>\n",
       "      <td>-6.54</td>\n",
       "      <td>38.7</td>\n",
       "      <td>16.2</td>\n",
       "      <td>36.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TARA_004_SRF_0.22-1.6</td>\n",
       "      <td>0.071865</td>\n",
       "      <td>0.044459</td>\n",
       "      <td>0.004822</td>\n",
       "      <td>0.030419</td>\n",
       "      <td>0.016433</td>\n",
       "      <td>0.004164</td>\n",
       "      <td>0.131823</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.040992</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>36.55</td>\n",
       "      <td>-6.57</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.5</td>\n",
       "      <td>36.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TARA_007_DCM_0.22-1.6</td>\n",
       "      <td>0.054858</td>\n",
       "      <td>0.085269</td>\n",
       "      <td>0.002647</td>\n",
       "      <td>0.029133</td>\n",
       "      <td>0.018208</td>\n",
       "      <td>0.003917</td>\n",
       "      <td>0.062948</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.031687</td>\n",
       "      <td>...</td>\n",
       "      <td>(MEDI) Mediterranean Sea, Black Sea Province [...</td>\n",
       "      <td>37.04</td>\n",
       "      <td>1.95</td>\n",
       "      <td>41.7</td>\n",
       "      <td>17.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TARA_007_SRF_0.22-1.6</td>\n",
       "      <td>0.065118</td>\n",
       "      <td>0.094020</td>\n",
       "      <td>0.011056</td>\n",
       "      <td>0.051516</td>\n",
       "      <td>0.030656</td>\n",
       "      <td>0.001844</td>\n",
       "      <td>0.048174</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.037236</td>\n",
       "      <td>...</td>\n",
       "      <td>(MEDI) Mediterranean Sea, Black Sea Province [...</td>\n",
       "      <td>37.02</td>\n",
       "      <td>1.95</td>\n",
       "      <td>7.5</td>\n",
       "      <td>23.8</td>\n",
       "      <td>37.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TARA_009_DCM_0.22-1.6</td>\n",
       "      <td>0.071004</td>\n",
       "      <td>0.057010</td>\n",
       "      <td>0.003618</td>\n",
       "      <td>0.023745</td>\n",
       "      <td>0.030761</td>\n",
       "      <td>0.007429</td>\n",
       "      <td>0.042233</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>0.006716</td>\n",
       "      <td>...</td>\n",
       "      <td>(MEDI) Mediterranean Sea, Black Sea Province [...</td>\n",
       "      <td>39.07</td>\n",
       "      <td>5.86</td>\n",
       "      <td>55.0</td>\n",
       "      <td>16.2</td>\n",
       "      <td>37.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>TARA_151_DCM_0.22-3</td>\n",
       "      <td>0.072100</td>\n",
       "      <td>0.056772</td>\n",
       "      <td>0.004461</td>\n",
       "      <td>0.023119</td>\n",
       "      <td>0.038456</td>\n",
       "      <td>0.010471</td>\n",
       "      <td>0.038078</td>\n",
       "      <td>0.000588</td>\n",
       "      <td>0.007151</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>36.19</td>\n",
       "      <td>-28.88</td>\n",
       "      <td>77.6</td>\n",
       "      <td>16.8</td>\n",
       "      <td>36.2</td>\n",
       "      <td>228.5</td>\n",
       "      <td>1.6</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>TARA_151_SRF_0.22-3</td>\n",
       "      <td>0.083244</td>\n",
       "      <td>0.059927</td>\n",
       "      <td>0.004423</td>\n",
       "      <td>0.041899</td>\n",
       "      <td>0.013293</td>\n",
       "      <td>0.007387</td>\n",
       "      <td>0.081571</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.020850</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>36.16</td>\n",
       "      <td>-29.01</td>\n",
       "      <td>5.4</td>\n",
       "      <td>17.3</td>\n",
       "      <td>36.2</td>\n",
       "      <td>232.1</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>TARA_152_MES_0.22-3</td>\n",
       "      <td>0.064621</td>\n",
       "      <td>0.040111</td>\n",
       "      <td>0.002454</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>0.066670</td>\n",
       "      <td>0.001110</td>\n",
       "      <td>0.030238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006767</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>43.74</td>\n",
       "      <td>-16.88</td>\n",
       "      <td>791.8</td>\n",
       "      <td>10.2</td>\n",
       "      <td>35.5</td>\n",
       "      <td>175.3</td>\n",
       "      <td>18.2</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>TARA_152_MIX_0.22-3</td>\n",
       "      <td>0.089275</td>\n",
       "      <td>0.059603</td>\n",
       "      <td>0.003870</td>\n",
       "      <td>0.011616</td>\n",
       "      <td>0.025056</td>\n",
       "      <td>0.010832</td>\n",
       "      <td>0.049024</td>\n",
       "      <td>0.000550</td>\n",
       "      <td>0.009690</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>43.79</td>\n",
       "      <td>-16.89</td>\n",
       "      <td>23.9</td>\n",
       "      <td>14.3</td>\n",
       "      <td>36.0</td>\n",
       "      <td>239.4</td>\n",
       "      <td>3.8</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>TARA_152_SRF_0.22-3</td>\n",
       "      <td>0.090463</td>\n",
       "      <td>0.062083</td>\n",
       "      <td>0.004455</td>\n",
       "      <td>0.012197</td>\n",
       "      <td>0.020034</td>\n",
       "      <td>0.013354</td>\n",
       "      <td>0.051405</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.011858</td>\n",
       "      <td>...</td>\n",
       "      <td>(NAST-E) North Atlantic Subtropical Gyral Prov...</td>\n",
       "      <td>43.69</td>\n",
       "      <td>-16.85</td>\n",
       "      <td>5.4</td>\n",
       "      <td>14.3</td>\n",
       "      <td>36.0</td>\n",
       "      <td>243.1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     label  Amino acid metabolism and transport  \\\n",
       "0    TARA_004_DCM_0.22-1.6                             0.083391   \n",
       "1    TARA_004_SRF_0.22-1.6                             0.071865   \n",
       "2    TARA_007_DCM_0.22-1.6                             0.054858   \n",
       "3    TARA_007_SRF_0.22-1.6                             0.065118   \n",
       "4    TARA_009_DCM_0.22-1.6                             0.071004   \n",
       "..                     ...                                  ...   \n",
       "238    TARA_151_DCM_0.22-3                             0.072100   \n",
       "239    TARA_151_SRF_0.22-3                             0.083244   \n",
       "240    TARA_152_MES_0.22-3                             0.064621   \n",
       "241    TARA_152_MIX_0.22-3                             0.089275   \n",
       "242    TARA_152_SRF_0.22-3                             0.090463   \n",
       "\n",
       "     Carbohydrate metabolism and transport  Cell cycle control and mitosis  \\\n",
       "0                                 0.062391                        0.004614   \n",
       "1                                 0.044459                        0.004822   \n",
       "2                                 0.085269                        0.002647   \n",
       "3                                 0.094020                        0.011056   \n",
       "4                                 0.057010                        0.003618   \n",
       "..                                     ...                             ...   \n",
       "238                               0.056772                        0.004461   \n",
       "239                               0.059927                        0.004423   \n",
       "240                               0.040111                        0.002454   \n",
       "241                               0.059603                        0.003870   \n",
       "242                               0.062083                        0.004455   \n",
       "\n",
       "     Cell motility  Cell wall/membrane/envelope biogenesis  \\\n",
       "0         0.020305                                0.019107   \n",
       "1         0.030419                                0.016433   \n",
       "2         0.029133                                0.018208   \n",
       "3         0.051516                                0.030656   \n",
       "4         0.023745                                0.030761   \n",
       "..             ...                                     ...   \n",
       "238       0.023119                                0.038456   \n",
       "239       0.041899                                0.013293   \n",
       "240       0.006903                                0.066670   \n",
       "241       0.011616                                0.025056   \n",
       "242       0.012197                                0.020034   \n",
       "\n",
       "     Chromatin structure and dynamics  Coenzyme metabolism  Cytoskeleton  \\\n",
       "0                            0.011587             0.076737      0.000116   \n",
       "1                            0.004164             0.131823      0.000523   \n",
       "2                            0.003917             0.062948      0.000018   \n",
       "3                            0.001844             0.048174      0.000017   \n",
       "4                            0.007429             0.042233      0.000152   \n",
       "..                                ...                  ...           ...   \n",
       "238                          0.010471             0.038078      0.000588   \n",
       "239                          0.007387             0.081571      0.000408   \n",
       "240                          0.001110             0.030238      0.000000   \n",
       "241                          0.010832             0.049024      0.000550   \n",
       "242                          0.013354             0.051405      0.000779   \n",
       "\n",
       "     Defense mechanisms  ...  \\\n",
       "0              0.010855  ...   \n",
       "1              0.040992  ...   \n",
       "2              0.031687  ...   \n",
       "3              0.037236  ...   \n",
       "4              0.006716  ...   \n",
       "..                  ...  ...   \n",
       "238            0.007151  ...   \n",
       "239            0.020850  ...   \n",
       "240            0.006767  ...   \n",
       "241            0.009690  ...   \n",
       "242            0.011858  ...   \n",
       "\n",
       "                                              Province  Latitude  Longitude  \\\n",
       "0    (NAST-E) North Atlantic Subtropical Gyral Prov...     36.57      -6.54   \n",
       "1    (NAST-E) North Atlantic Subtropical Gyral Prov...     36.55      -6.57   \n",
       "2    (MEDI) Mediterranean Sea, Black Sea Province [...     37.04       1.95   \n",
       "3    (MEDI) Mediterranean Sea, Black Sea Province [...     37.02       1.95   \n",
       "4    (MEDI) Mediterranean Sea, Black Sea Province [...     39.07       5.86   \n",
       "..                                                 ...       ...        ...   \n",
       "238  (NAST-E) North Atlantic Subtropical Gyral Prov...     36.19     -28.88   \n",
       "239  (NAST-E) North Atlantic Subtropical Gyral Prov...     36.16     -29.01   \n",
       "240  (NAST-E) North Atlantic Subtropical Gyral Prov...     43.74     -16.88   \n",
       "241  (NAST-E) North Atlantic Subtropical Gyral Prov...     43.79     -16.89   \n",
       "242  (NAST-E) North Atlantic Subtropical Gyral Prov...     43.69     -16.85   \n",
       "\n",
       "     Depth (m)  Temperature (°C)  Salinity (PSU)  Oxygen (µmol/kg)  \\\n",
       "0         38.7              16.2            36.6               NaN   \n",
       "1         10.0              20.5            36.6               NaN   \n",
       "2         41.7              17.4             NaN               NaN   \n",
       "3          7.5              23.8            37.5               NaN   \n",
       "4         55.0              16.2            37.8               NaN   \n",
       "..         ...               ...             ...               ...   \n",
       "238       77.6              16.8            36.2             228.5   \n",
       "239        5.4              17.3            36.2             232.1   \n",
       "240      791.8              10.2            35.5             175.3   \n",
       "241       23.9              14.3            36.0             239.4   \n",
       "242        5.4              14.3            36.0             243.1   \n",
       "\n",
       "     Nitrates (µmol/L)  NO2 (µmol/L)  PO4 (µmol/L)  \n",
       "0                  NaN           NaN           NaN  \n",
       "1                  NaN           NaN           NaN  \n",
       "2                  NaN          0.00          0.01  \n",
       "3                  NaN          0.00          0.06  \n",
       "4                  NaN          0.02          0.02  \n",
       "..                 ...           ...           ...  \n",
       "238                1.6          0.01          0.01  \n",
       "239                0.3          0.02          0.01  \n",
       "240               18.2          0.01          1.12  \n",
       "241                3.8          0.32          0.17  \n",
       "242                3.0          0.31          0.16  \n",
       "\n",
       "[243 rows x 36 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OG_relative_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6ba3768e-da2e-483b-bbcd-37e193170caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "OG_relative_df = OG_relative_df.replace(['(DCM) deep chlorophyll maximum layer (ENVO:01000326)', \n",
    "                 '(SRF) surface water layer (ENVO:00002042)',\n",
    "                 '(MES) mesopelagic zone (ENVO:00000213) & marine oxygen minimum zone (ENVO:01000065)',\n",
    "                 '(MES) mesopelagic zone (ENVO:00000213)',\n",
    "                 '(DCM) deep chlorophyll maximum layer (ENVO:01000326) & marine oxygen minimum zone (ENVO:01000065)',\n",
    "                 '(MIX) marine epipelagic mixed layer (ENVO:01000061)',\n",
    "                 'Westerlies Biome',\n",
    "                 'Coastal Biome ',\n",
    "                 'Trades Biome',\n",
    "                 'Polar Biome',\n",
    "                 '(NAO) North Atlantic Ocean [MRGID:1912]',\n",
    "                 '(MS) Mediterranean Sea [MRGID:1905]', \n",
    "                 '(RS) Red Sea [MRGID:4264]',\n",
    "                 '(IO) Indian Ocean [MRGID:1904]',\n",
    "                 '(SAO) South Atlantic Ocean [MRGID:1914]',\n",
    "                 '(SO) Southern Ocean [MRGID:1907]',\n",
    "                 '(SPO) South Pacific Ocean [MRGID:1910]',\n",
    "                 '(NPO) North Pacific Ocean [MRGID:1908]',\n",
    "                 '(NAST-E) North Atlantic Subtropical Gyral Province [MRGID:21467]',\n",
    "                 '(MEDI) Mediterranean Sea, Black Sea Province [MRGID:21465]',\n",
    "                 '(REDS) Red Sea, Persian Gulf Province [MRGID:21474]',\n",
    "                 '(ARAB) Northwest Arabian Sea Upwelling Province [MRGID:21475]',\n",
    "                 '(MONS) Indian Monsoon Gyres Province [MRGID:21471]',\n",
    "                 '(ISSG) Indian South Subtropical Gyre Province [MRGID:21472]',\n",
    "                 '(EAFR) Eastern Africa Coastal Province [MRGID:21473]',\n",
    "                 '(BENG) Benguela Current Coastal Province [MRGID:21470]',\n",
    "                 '(SATL) South Atlantic Gyral Province [MRGID:21459]',\n",
    "                 '(FKLD) Southwest Atlantic Shelves Province [MRGID:21469]',\n",
    "                 '(ANTA) Antarctic Province [MRGID:21502]',\n",
    "                 '(CHIL) Chile-Peru Current Coastal Province [MRGID:21495]',\n",
    "                 '(SPSG) South Pacific Subtropical Gyre Province, North and South [MRGID:21486]',\n",
    "                 '(PEOD) Pacific Equatorial Divergence Province [MRGID:21489]',\n",
    "                 '(NPST) North Pacific Subtropical and Polar Front Provinces [MRGID:21484]',\n",
    "                 '(PNEC) North Pacific Equatorial Countercurrent Province [MRGID:21488]',\n",
    "                 '(CAMR) Central American Coastal Province [MRGID:21494]',\n",
    "                 '(GUIA) Guianas Coastal Province [MRGID:21463]',\n",
    "                 '(CARB) Caribbean Province [MRGID:21466]',\n",
    "                 '(GFST) Gulf Stream Province [MRGID:21454]',\n",
    "                 '(NAST-W) North Atlantic Subtropical Gyral Province [MRGID:21455]'\n",
    "                ], \n",
    "                ['DCM', 'SRF', 'MES/OMZ', 'MES', 'DCM/OMZ', 'MIX', 'Westerlies', 'Coastal',\n",
    "                 'Trades', 'Polar', 'NAO', 'MS',  'RS', 'IO', 'SAO', 'SO', 'SPO', 'NPO', 'NAST-E',\n",
    "                 'MEDI', 'REDS', 'ARAB', 'MONS', 'ISSG', 'EAFR', 'BENG', 'SATL', 'FKLD', 'ANTA', 'CHIL',\n",
    "                 'SPSG', 'PEOD', 'NPST', 'PNEC', 'CAMR', 'GUIA', 'CARB', 'GFST', 'NAST-W'\n",
    "                ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7fdde2fd-a3c7-41c6-89f2-49222ecf16c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "relative_known_function_file = os.path.join(data_path, \"relative_known_function_with_metadata.csv\")\n",
    "OG_relative_df.to_csv(relative_known_function_file, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
